File: core/src/main/java/io/snabble/sdk/checkout/DefaultCheckoutApi.java
package io.snabble.sdk.checkout;

import androidx.annotation.NonNull;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import io.snabble.sdk.PaymentMethod;
import io.snabble.sdk.Product;
import io.snabble.sdk.Project;
import io.snabble.sdk.ShoppingCart;
import io.snabble.sdk.Snabble;
import io.snabble.sdk.payment.PaymentCredentials;
import io.snabble.sdk.utils.DateUtils;
import io.snabble.sdk.utils.GsonHolder;
import io.snabble.sdk.utils.JsonCallback;
import io.snabble.sdk.utils.Logger;
import io.snabble.sdk.utils.SimpleJsonCallback;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class DefaultCheckoutApi implements CheckoutApi {
    private static final MediaType JSON = MediaType.parse("application/json");

    private final Project project;
    private final ShoppingCart shoppingCart;
    private final OkHttpClient okHttpClient;
    private Call call;

    public DefaultCheckoutApi(Project project, ShoppingCart shoppingCart) {
        this.project = project;
        this.shoppingCart = shoppingCart;
        this.okHttpClient = project.getOkHttpClient();
    }

    @Override
    public void cancel() {
        if (call != null) {
            call.cancel();
            call = null;
        }
    }

    @Override
    public void abort(final CheckoutProcessResponse checkoutProcessResponse, final PaymentAbortResult paymentAbortResult) {
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(checkoutProcessResponse.getSelfLink()))
                .patch(RequestBody.create(JSON, "{\"aborted\":true}"))
                .build();

        okHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response.isSuccessful()) {
                    Logger.d("Payment aborted");
                    if (paymentAbortResult != null) {
                        paymentAbortResult.success();
                    }
                } else {
                    Logger.e("Error while aborting payment");
                    if (paymentAbortResult != null) {
                        paymentAbortResult.error();
                    }
                }
            }

            @Override
            public void onFailure(Call call, IOException e) {
                Logger.e("Error while aborting payment");
                if (paymentAbortResult != null) {
                    paymentAbortResult.error();
                }
            }
        });
    }

    @Override
    public void createCheckoutInfo(final ShoppingCart.BackendCart backendCart,
                                   final List<? extends PaymentMethod> clientAcceptedPaymentMethods,
                                   final CheckoutInfoResult checkoutInfoResult,
                                   final long timeout) {
        String checkoutUrl = project.getCheckoutUrl();
        if (checkoutUrl == null) {
            Logger.e("Could not checkout, no checkout url provided in metadata");
            checkoutInfoResult.connectionError();
            return;
        }

        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(checkoutUrl))
                .post(RequestBody.create(JSON, GsonHolder.get().toJson(backendCart)))
                .build();

        cancel();

        OkHttpClient okClient = okHttpClient;

        if (timeout > 0) {
            okClient = okClient.newBuilder()
                    .callTimeout(timeout, TimeUnit.MILLISECONDS)
                    .build();
        }

        call = okClient.newCall(request);
        call.enqueue(new JsonCallback<SignedCheckoutInfo, JsonObject>(SignedCheckoutInfo.class, JsonObject.class) {
            @Override
            public void success(SignedCheckoutInfo signedCheckoutInfo) {
                int price;

                if (signedCheckoutInfo.getCheckoutInfo().has("price")
                        && signedCheckoutInfo.getCheckoutInfo().get("price").getAsJsonObject().has("price")) {
                    price = signedCheckoutInfo.getCheckoutInfo()
                            .get("price")
                            .getAsJsonObject()
                            .get("price")
                            .getAsInt();
                } else {
                    price = shoppingCart.getTotalPrice();
                }

                List<PaymentMethodInfo> availablePaymentMethods = signedCheckoutInfo.getAvailablePaymentMethods(clientAcceptedPaymentMethods);
                if (!availablePaymentMethods.isEmpty()) {
                    checkoutInfoResult.success(signedCheckoutInfo, price, availablePaymentMethods);
                } else {
                    checkoutInfoResult.connectionError();
                }
            }

            @Override
            public void failure(JsonObject obj) {
                try {
                    JsonObject error = obj.get("error").getAsJsonObject();
                    String type = error.get("type").getAsString();

                    switch (type) {
                        case "invalid_cart_item":
                            List<String> invalidSkus = new ArrayList<>();
                            JsonArray arr = error.get("details").getAsJsonArray();

                            for (int i=0; i<arr.size(); i++) {
                                String sku = arr.get(0).getAsJsonObject().get("sku").getAsString();
                                invalidSkus.add(sku);
                            }

                            List<Product> invalidProducts = new ArrayList<>();
                            for (int i=0; i<shoppingCart.size(); i++) {
                                Product product = shoppingCart.get(i).getProduct();
                                if (product != null) {
                                    if (invalidSkus.contains(product.getSku())) {
                                        invalidProducts.add(product);
                                    }
                                }
                            }

                            Logger.e("Invalid products");
                            checkoutInfoResult.invalidProducts(invalidProducts);
                            break;
                        case "no_available_method":
                            checkoutInfoResult.noAvailablePaymentMethod();
                            break;
                        case "bad_shop_id":
                        case "shop_not_found":
                            checkoutInfoResult.noShop();
                            break;
                        case "invalid_deposit_return_voucher":
                            checkoutInfoResult.invalidDepositReturnVoucher();
                            break;
                        default:
                            checkoutInfoResult.unknownError();
                            break;
                    }
                } catch (Exception e) {
                    error(e);
                }
            }

            @Override
            public void error(Throwable t) {
                Logger.e("Error creating checkout info: " + t.getMessage());
                checkoutInfoResult.connectionError();
            }
        });
    }

    @Override
    public void updatePaymentProcess(final String url,
                                     final PaymentProcessResult paymentProcessResult) {
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(url))
                .get()
                .build();

        cancel();

        call = okHttpClient.newCall(request);
        call.enqueue(new SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse.class) {
            @Override
            public void success(CheckoutProcessResponse checkoutProcessResponse) {
                paymentProcessResult.success(checkoutProcessResponse, rawResponse());
                DefaultCheckoutApi.this.call = null;
            }

            @Override
            public void error(Throwable t) {
                paymentProcessResult.error();
            }
        });
    }

    @Override
    public void updatePaymentProcess(final CheckoutProcessResponse checkoutProcessResponse,
                                     final PaymentProcessResult paymentProcessResult) {
        String url = checkoutProcessResponse.getSelfLink();
        if (url == null) {
            paymentProcessResult.error();
            return;
        }

        updatePaymentProcess(url, paymentProcessResult);
    }

    @Override
    public void createPaymentProcess(final String id,
                                     final SignedCheckoutInfo signedCheckoutInfo,
                                     final PaymentMethod paymentMethod,
                                     final PaymentCredentials paymentCredentials,
                                     final boolean processedOffline,
                                     final Date finalizedAt,
                                     final PaymentProcessResult paymentProcessResult) {
        CheckoutProcessRequest checkoutProcessRequest = new CheckoutProcessRequest();
        checkoutProcessRequest.setPaymentMethod(paymentMethod);
        checkoutProcessRequest.setSignedCheckoutInfo(signedCheckoutInfo);

        if (processedOffline) {
            checkoutProcessRequest.setProcessedOffline(true);

            if (finalizedAt != null) {
                checkoutProcessRequest.setFinalizedAt(DateUtils.toRFC3339(finalizedAt));
            }
        }

        if (paymentCredentials != null) {
            checkoutProcessRequest.setPaymentInformation(new PaymentInformation());

            String data = paymentCredentials.getEncryptedData();
            if (data == null) {
                paymentProcessResult.error();
                return;
            }

            checkoutProcessRequest.getPaymentInformation().setOriginType(paymentCredentials.getType().getOriginType());
            checkoutProcessRequest.getPaymentInformation().setEncryptedOrigin(data);

            if (paymentCredentials.getType() == PaymentCredentials.Type.CREDIT_CARD_PSD2) {
                Date date = new Date(paymentCredentials.getValidTo());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy/MM/dd");
                checkoutProcessRequest.getPaymentInformation().setValidUntil(simpleDateFormat.format(date));
                checkoutProcessRequest.getPaymentInformation().setCardNumber(paymentCredentials.getObfuscatedId());
            } else if (paymentCredentials.getType() == PaymentCredentials.Type.PAYDIREKT) {
                Map<String, String> additionalData = paymentCredentials.getAdditionalData();
                if (additionalData != null) {
                    checkoutProcessRequest.getPaymentInformation().setDeviceID(additionalData.get("deviceID"));
                    checkoutProcessRequest.getPaymentInformation().setDeviceName(additionalData.get("deviceName"));
                    checkoutProcessRequest.getPaymentInformation().setDeviceFingerprint(additionalData.get("deviceFingerprint"));
                    checkoutProcessRequest.getPaymentInformation().setDeviceIPAddress(additionalData.get("deviceIPAddress"));
                }
            }
        }

        String url = signedCheckoutInfo.getCheckoutProcessLink();
        if (url == null) {
            paymentProcessResult.error();
            return;
        }

        url = url + "/" + id;
        String json = GsonHolder.get().toJson(checkoutProcessRequest);
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(url))
                .put(RequestBody.create(JSON, json))
                .build();

        cancel();

        call = okHttpClient.newCall(request);

        String finalUrl = url;
        call.enqueue(new SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse.class) {
            @Override
            public void success(CheckoutProcessResponse checkoutProcess) {
                paymentProcessResult.success(checkoutProcess, rawResponse());
            }

            @Override
            public void error(Throwable t) {
                if (responseCode() == 403) {
                    updatePaymentProcess(finalUrl, paymentProcessResult);
                } else {
                    paymentProcessResult.error();
                }
            }
        });
    }

    @Override
    public void authorizePayment(final CheckoutProcessResponse checkoutProcessResponse,
                                 final AuthorizePaymentRequest authorizePaymentRequest,
                                 final AuthorizePaymentResult authorizePaymentResult) {
        String url = checkoutProcessResponse.getAuthorizePaymentLink();
        if (url == null) {
            authorizePaymentResult.error();
            return;
        }

        String json = GsonHolder.get().toJson(authorizePaymentRequest);
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(url))
                .post(RequestBody.create(JSON, json))
                .build();

        Call call = okHttpClient.newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                authorizePaymentResult.error();
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) {
                if (response.isSuccessful()) {
                    authorizePaymentResult.success();
                } else {
                    authorizePaymentResult.error();
                }
            }
        });
    }
}

File: core/src/main/java/io/snabble/sdk/checkout/DefaultCheckoutApi.kt
package io.snabble.sdk.checkout

import android.annotation.SuppressLint
import com.google.gson.JsonObject
import io.snabble.sdk.*
import io.snabble.sdk.payment.PaymentCredentials
import io.snabble.sdk.shoppingcart.ShoppingCart
import io.snabble.sdk.shoppingcart.data.cart.BackendCart
import io.snabble.sdk.utils.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.net.HttpURLConnection.HTTP_CONFLICT
import java.net.HttpURLConnection.HTTP_FORBIDDEN
import java.net.HttpURLConnection.HTTP_NOT_FOUND
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

class DefaultCheckoutApi(private val project: Project,
                         private val shoppingCart: ShoppingCart
) : CheckoutApi {

    private val okHttpClient: OkHttpClient = project.okHttpClient.newBuilder()
        .addInterceptor(HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BODY })
        .build()

    private var call: Call? = null

    override fun cancel() {
        call?.cancel()
        call = null
    }

    override fun abort(
        checkoutProcessResponse: CheckoutProcessResponse,
        paymentAbortResult: PaymentAbortResult?
    ) {
        if (checkoutProcessResponse.selfLink == null) {
            paymentAbortResult?.onError()
            return
        }

        val request = Request.Builder()
            .url(Snabble.absoluteUrl(checkoutProcessResponse.selfLink!!))
            .patch("{\"aborted\":true}".toRequestBody(JSON))
            .build()

        okHttpClient.newCall(request).enqueue(object : Callback {
            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    Logger.d("Payment aborted")
                    paymentAbortResult?.onSuccess()
                } else {
                    Logger.e("Error while aborting payment")
                    paymentAbortResult?.onError()
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                Logger.e("Error while aborting payment")
                paymentAbortResult?.onError()
            }
        })
    }

    override fun createCheckoutInfo(
        backendCart: BackendCart,
        checkoutInfoResult: CheckoutInfoResult?,
        timeout: Long
    ) {
        val checkoutUrl = project.checkoutUrl
        if (checkoutUrl == null) {
            Logger.e("Could not checkout, no checkout url provided in metadata")
            checkoutInfoResult?.onConnectionError()
            return
        }

        val request = Request.Builder()
            .url(Snabble.absoluteUrl(checkoutUrl))
            .post(GsonHolder.get().toJson(backendCart).toRequestBody(JSON))
            .build()

        cancel()

        var okClient = okHttpClient
        if (timeout > 0) {
            okClient = okClient.newBuilder()
                .callTimeout(timeout, TimeUnit.MILLISECONDS)
                .build()
        }

        call = okClient.newCall(request)
        call?.enqueue(object : JsonCallback<SignedCheckoutInfo, JsonObject>(
            SignedCheckoutInfo::class.java, JsonObject::class.java) {
            override fun success(signedCheckoutInfo: SignedCheckoutInfo) {
                val price = signedCheckoutInfo.checkoutInfo
                    ?.get("price")
                    ?.asJsonObject
                    ?.get("price")
                    ?.asInt
                    ?: shoppingCart.totalPrice

                val availablePaymentMethods = signedCheckoutInfo.getAvailablePaymentMethods()
                checkoutInfoResult?.onSuccess(signedCheckoutInfo, price, availablePaymentMethods)
            }

            override fun failure(obj: JsonObject) {
                try {
                    val error: JsonObject? = obj["error"].asJsonObject
                    val type: String? = error?.get("type")?.asString
                    when (type) {
                        "invalid_cart_item" -> {
                            val invalidSkus = error["details"].asJsonArray
                                .mapNotNull { it.asJsonObject["sku"].asString?.ifBlank { null } }

                            val invalidProducts = shoppingCart
                                .mapNotNull { it?.product }
                                .filter { invalidSkus.contains(it.sku) }

                            if (invalidProducts.isNotEmpty()) {
                                Logger.e("Invalid products found")
                                checkoutInfoResult?.onInvalidProducts(invalidProducts)
                            } else {
                                Logger.e("Invalid items found")
                                val invalidIds = error["details"].asJsonArray
                                    .mapNotNull { it.asJsonObject["id"].asString }
                                checkoutInfoResult?.onInvalidItems(invalidIds)
                            }
                        }

                        "no_available_method" -> checkoutInfoResult?.onNoAvailablePaymentMethodFound()
                        "bad_shop_id", "shop_not_found" -> checkoutInfoResult?.onNoShopFound()
                        else -> checkoutInfoResult?.onUnknownError()
                    }
                } catch (e: IllegalStateException) {
                    error(e)
                } catch (e: UnsupportedOperationException) {
                    error(e)
                }
            }

            override fun error(t: Throwable) {
                Logger.e("Error creating checkout info: " + t.message)
                checkoutInfoResult?.onConnectionError()
            }
        })
    }

    private fun updatePaymentProcess(
        url: String,
        paymentProcessResult: PaymentProcessResult?
    ) {
        val request = Request.Builder()
            .url(Snabble.absoluteUrl(url))
            .get()
            .build()

        cancel()

        call = okHttpClient.newCall(request)
        call?.enqueue(object : SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse::class.java) {
            override fun success(checkoutProcessResponse: CheckoutProcessResponse) {
                paymentProcessResult?.onSuccess(checkoutProcessResponse, rawResponse())
                call = null
            }

            override fun error(t: Throwable) {
                if (responseCode() == HTTP_NOT_FOUND) {
                    paymentProcessResult?.onNotFound()
                } else {
                    paymentProcessResult?.onError()
                }
            }
        })
    }

    override fun updatePaymentProcess(
        checkoutProcessResponse: CheckoutProcessResponse,
        paymentProcessResult: PaymentProcessResult?
    ) {
        val url = checkoutProcessResponse.selfLink
        if (url == null) {
            paymentProcessResult?.onError()
            return
        }
        updatePaymentProcess(url, paymentProcessResult)
    }

    @SuppressLint("SimpleDateFormat")
    override fun createPaymentProcess(
        id: String,
        signedCheckoutInfo: SignedCheckoutInfo,
        paymentMethod: PaymentMethod,
        processedOffline: Boolean,
        paymentCredentials: PaymentCredentials?,
        finalizedAt: Date?,
        paymentProcessResult: PaymentProcessResult?
    ) {
        if (paymentCredentials != null) {
            val data = paymentCredentials.encryptedData
            if (data == null) {
                paymentProcessResult?.onError()
                return
            }
        }

        val originType = paymentCredentials?.type?.originType
            ?: getOriginTypeFromPaymentMethodDescriptor(paymentMethod = paymentMethod)

        val checkoutProcessRequest = CheckoutProcessRequest(
            paymentMethod = paymentMethod,
            signedCheckoutInfo = signedCheckoutInfo,
            processedOffline = processedOffline,
            finalizedAt = finalizedAt?.let { DateUtils.toRFC3339(it) },
            paymentInformation = when (paymentCredentials?.type) {
                PaymentCredentials.Type.EXTERNAL_BILLING -> {
                    PaymentInformation(
                        originType = originType,
                        encryptedOrigin = paymentCredentials.encryptedData,
                        subject = paymentCredentials.additionalData["subject"]
                    )
                }

                PaymentCredentials.Type.CREDIT_CARD_PSD2 -> {
                    PaymentInformation(
                        originType = originType,
                        encryptedOrigin = paymentCredentials.encryptedData,
                        validUntil = SimpleDateFormat("yyyy/MM/dd").format(Date(paymentCredentials.validTo)),
                        cardNumber = paymentCredentials.obfuscatedId,
                    )
                }

                PaymentCredentials.Type.GIROPAY -> {
                    PaymentInformation(
                        originType = originType,
                        encryptedOrigin = paymentCredentials.encryptedData,
                        deviceID = paymentCredentials.additionalData["deviceID"],
                        deviceName = paymentCredentials.additionalData["deviceName"],
                        deviceFingerprint = paymentCredentials.additionalData["deviceFingerprint"],
                        deviceIPAddress = paymentCredentials.additionalData["deviceIPAddress"],
                    )
                }
                null -> null
                else -> {
                    PaymentInformation(
                        originType = originType,
                        encryptedOrigin = paymentCredentials.encryptedData
                    )
                }
            }
        )

        var url = signedCheckoutInfo.checkoutProcessLink
        if (url == null) {
            paymentProcessResult?.onError()
            return
        }
        url = "$url/$id"

        val json = GsonHolder.get().toJson(checkoutProcessRequest)
        val request = Request.Builder()
            .url(Snabble.absoluteUrl(url))
            .put(json.toRequestBody(JSON))
            .build()

        cancel()

        call = okHttpClient.newCall(request)
        call?.enqueue(object :
            SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse::class.java) {
            override fun success(checkoutProcess: CheckoutProcessResponse) {
                paymentProcessResult?.onSuccess(checkoutProcess, rawResponse())
            }

            override fun error(t: Throwable) {
                // Legacy: In case of a conflicting checkout process the backend currently responds
                // with a 403 (FORBIDDEN).
                // This will change in the future to the correct status 409 (CONFLICT).
                if (responseCode() == HTTP_FORBIDDEN || responseCode() == HTTP_CONFLICT) {
                    updatePaymentProcess(url, paymentProcessResult)
                } else if (responseCode() == HTTP_NOT_FOUND) {
                    paymentProcessResult?.onNotFound()
                } else {
                    paymentProcessResult?.onError()
                }
            }
        })
    }

    override fun authorizePayment(
        checkoutProcessResponse: CheckoutProcessResponse,
        authorizePaymentRequest: AuthorizePaymentRequest,
        authorizePaymentResult: AuthorizePaymentResult?
    ) {
        val url = checkoutProcessResponse.authorizePaymentLink
        if (url == null) {
            authorizePaymentResult?.onError()
            return
        }

        val json = GsonHolder.get().toJson(authorizePaymentRequest)
        val request = Request.Builder()
            .url(Snabble.absoluteUrl(url))
            .post(json.toRequestBody(JSON))
            .build()

        val call = okHttpClient.newCall(request)
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                authorizePaymentResult?.onError()
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    authorizePaymentResult?.onSuccess()
                } else {
                    authorizePaymentResult?.onError()
                }
            }
        })
    }

    private fun getOriginTypeFromPaymentMethodDescriptor(paymentMethod: PaymentMethod): String? =
        project.paymentMethodDescriptors
            .firstOrNull { it.paymentMethod == paymentMethod }
            ?.acceptedOriginTypes
            ?.get(0)

    companion object {

        private val JSON: MediaType = "application/json".toMediaType()
    }
}
